Update the checkpointing guide with Orbax migration code examples (#2688)
google/flax: docs/guides/use_checkpointing.md

---
jupyter:
  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.8
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
---

# Save and load checkpoints

> In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints) — a simplistic and generic checkpointing library built into Flax. Flax allows you to save and load model parameters, metadata, and a variety of Python data using the library [Orbax](https://github.com/google/orbax), which provides a more customizable and flexible API for various array types and storage formats. In addition, Flax provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.

Reviewer: Since checkpointing will be taken care of by Orbax, maybe we can omit `flax.training.checkpoints` now.

For example:

```suggestion
In this guide, you will learn about saving and loading checkpoints with Flax and Orbax](https://github.com/google/orbax). Flax allows you to save and load model parameters, metadata, and a variety of Python data using Orbax, which provides a more customizable and flexible API for various array types and storage formats. In addition, Flax provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.
```

Then, in the **Ongoing migration** note, we can add that:

```diff
**_Ongoing migration:_** In the foreseeable future, Flax's checkpointing
functionality will gradually be migrated to Orbax
+ from `flax.training,checkpoints`.

```

Does this make sense? LMKWYT :+1:

Author: Sounds good! Done.

> > **_Ongoing migration:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to Orbax from `flax.training,checkpoints`. All existed features in the Flax API will continue to be supported, though the API will change. You are encouraged to try out the new API by creating an `orbax.Checkpointer` and pass it in your Flax API calls as arg `orbax_checkpointer`. You may use this page as the most up-to-date code examples to use Orbax-Flax checkpointing.

Reviewer: Nit:

How about:

```suggestion
> **_Ongoing migration:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to Orbax from `flax.training.checkpoints`. All existing features in the Flax API will continue to be supported, but the API will change. You are encouraged to try out the new Orbax API by creating an `orbax.Checkpointer` and pass it in your Flax API calls as an argument `orbax_checkpointer`, as demonstrated later in this guide. This guide provides the most up-to-date code examples for using Orbax and Flax for checkpointing.
```

WDYT?

---
jupyter:
  jupytext:
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.8
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
---

> # Save / Load Checkpoints

Reviewer: Nit:

```suggestion
# Save / load checkpoints
```

or "Save and load..."

(with "sentence case" https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings)

> In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.

Reviewer: Suggestion:

- Maybe it's worth adding a sentence about Orbax - `orbax.Checkpointer` - which the user can use with `flax.checkpoints.save_checkpoint`, as demonstrated later the "Save checkpoints" section.
- Later, we can rephrase the entire introduction after the Orbax migration is complete.

For example:

```suggestion
In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.

In addition, you will be using the [Orbax](https://github.com/google/orbax) Checkpointer. Refer to the note about Orbax below.
```

> > **_Note:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to [Orbax](https://github.com/google/orbax), which provides a more customizable and flexible API for various array types and storage formats. All existed features of Flax will continue to be supported, though the API may be slightly changed. This page will contain the most up-to-date code examples to use Orbax-Flax checkpointing.

Reviewer: Suggestions:

- "All existing features in the Flax API will continue to be supported, though the API may change."
- "Once the migration is complete, this guide will contain the most up-to-date code examples with Orbax-Flax checkpointing."

```suggestion
> **_Note:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to [Orbax](https://github.com/google/orbax), which provides a more customizable and flexible API for various array types and storage formats. All existing features in the Flax API will continue to be supported, though the API may change. This page will contain the most up-to-date code examples to use Orbax-Flax checkpointing.
```

This guide covers the following:

* Basic saving and loading of checkpoints with [`flax.training.checkpoints.save_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint).
* More flexible and sustainable ways to load checkpoints ([`flax.training.checkpoints.restore_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.restore_checkpoint)).
* How to save and load checkpoints when you run in multi-host scenarios with
[`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess).

## Setup

Install/upgrade Flax, JAX, [Optax](https://optax.readthedocs.io/) and [TensorStore](https://google.github.io/tensorstore/). For JAX installation with GPU/TPU support, visit [this section on GitHub](https://github.com/google/jax#installation).

```python
!pip install -U -q flax jax jaxlib optax tensorstore orbax

> # Installing Flax and Orbax from head to use latest checkpoint features.
> # This part will be removed after both make a new release.

Reviewer: Suggestion:

```suggestion
# To use the latest checkpoint features, install Flax and Orbax from head.
# Note: This section will be removed/replaced after the new release for both libraries.
```

!pip3 install --upgrade --no-deps --force-reinstall \
"git+https://github.com/google/flax.git@main#egg=flax"
!pip3 install --upgrade --no-deps --force-reinstall \
"git+https://github.com/google/orbax.git@main#egg=orbax"
```

Note: Before running `import jax`, create eight fake devices to mimic multi-host checkpointing in this notebook. Note that the order of imports is important here. The `os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'` command works only with the CPU backend. This means it won't work with GPU/TPU acceleration on if you're running this notebook in Google Colab. If you are already running the code on multiple devices (for example, in a 4x2 TPU environment), you can skip running the next cell.

```python
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
```

```python
from typing import Optional, Any
import shutil

import numpy as np
import jax
from jax import random, numpy as jnp

import flax
from flax import linen as nn
from flax.training import checkpoints, train_state
from flax import struct, serialization
import orbax.checkpoint as orbax

import optax
```

```python

> # Orbax needs to enable asyncio in colab environment.

Reviewer: Nit:

```suggestion
# Orbax needs to enable asyncio in a Colab environment.
```

!pip install nest_asyncio
import nest_asyncio
nest_asyncio.apply()
```

## Save checkpoints

In Flax, you save and load any given JAX [pytree](https://jax.readthedocs.io/en/latest/pytrees.html) using the `flax.training.checkpoints` package. This includes not only typical Python and NumPy containers, but also customized classes extended from [`flax.struct.dataclass`](https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html#flax.struct.dataclass). That means you can store almost any data generated—not only your model parameters, but any arrays/dictionaries, metadata/configs, and so on.

Create a pytree with many data structures and containers, and play with it:

> ```python

Reviewer: This is cool. If the `jupytext` pre-commit test passes, we can manually change/improve all our other Markdown files.

# To use the latest checkpoint features, install Flax from head.
# Note: This part will be removed/replaced after the next Flax release.
!pip3 install --upgrade --no-deps --force-reinstall -qq \
"git+https://github.com/google/flax.git@main#egg=flax"
```

Note: Before running `import jax`, create eight fake devices to mimic multi-host checkpointing in this notebook. Note that the order of imports is important here. The `os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'` command works only with the CPU backend. This means it won't work with GPU/TPU acceleration on if you're running this notebook in Google Colab. If you are already running the code on multiple devices (for example, in a 4x2 TPU environment), you can skip running the next cell.

```python
import os
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
```

```python
from typing import Optional, Any
import shutil

import numpy as np
import jax
from jax import random, numpy as jnp

import flax
from flax import linen as nn
from flax.training import checkpoints, train_state
from flax import struct, serialization
import orbax.checkpoint as orbax

import optax
```

> ```python

Reviewer: If possible, maybe we can suggest this step after `pip install flax` since this is still in the "Setup" stage of the guide.

For example:

```
!pip install -U -qq flax orbax
...

# Orbax requires asyncio to be insabled in a Google Colab environment
!pip install nest_asyncio
```

Then in the next code cell:

```
import jax
...
import optax
...
import nest_asyncio
nest_asyncio.apply()
```

LMKWYT

* How to save and load checkpoints when you run in multi-host scenarios with
[`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess).

## Setup

Install/upgrade Flax, JAX, [Optax](https://optax.readthedocs.io/) and [TensorStore](https://google.github.io/tensorstore/). For JAX installation with GPU/TPU support, visit [this section on GitHub](https://github.com/google/jax#installation).

```python

> !pip install -U -qq flax jax jaxlib optax tensorstore orbax

Reviewer: If the `flax` package contains JAX, Optax, etc, are we instructing users to _then upgrade_ `jax`/`jaxlib` and `optax`? Otherwise, maybe `!pip install -U -q flax tensorstore orbax` might be enough?

Author: Done.

# Flax's TrainState is a pytree dataclass and is supported in checkpointing.
# Define your class with `@flax.struct.dataclass` decorator to make it compatible.
tx = optax.sgd(learning_rate=0.001)      # An Optax SGD optimizer.
state = train_state.TrainState.create(
    apply_fn=model.apply,
    params=variables['params'],
    tx=tx)

> # Make some simple fake grad update like the ones done in normal training.

Reviewer: Suggestion:

Since it's only a forward pass and the example in the guide is just a toy example, then maybe:

```suggestion
# Perform a simple gradient update similar to the one during a normal training workflow.
```

(Internally, `.apply_gradients` Flax and Optax (`tx`) code, isn't it?)

When saving a checkpoint, Flax will bookkeep the existing checkpoints based on your arguments. For example, by setting `overwrite=False` in [`flax.checkpoints.save_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint), Flax will not automatically save your checkpoint if there is already a step that is equal to or newer than the current one presently in the checkpoint directory. By setting `keep=2`, Flax will keep a maximum of 2 checkpoints in the directory. Learn more in the [API reference](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints).

> You can start to use Orbax to handle the underlying save by creating an `orbax.Checkpointer`, and pass it into the `flax.checkpoints.save_checkpoint` call.

Reviewer: If we haven't done this already, maybe we can include a link to `orbax.Checkpointer` API docs if they exist.

```python
# Import Flax Checkpoints.
from flax.training import checkpoints

ckpt_dir = 'tmp/flax-checkpointing'

if os.path.exists(ckpt_dir):
    shutil.rmtree(ckpt_dir)  # Remove any existing checkpoints from the last notebook run.

orbax_checkpointer = orbax.Checkpointer(orbax.PyTreeCheckpointHandler())
checkpoints.save_checkpoint(ckpt_dir=ckpt_dir,
                            target=ckpt,
                            step=0,
                            overwrite=False,
                            keep=2,
                            orbax_checkpointer=orbax_checkpointer)
```

## Restore checkpoints

To restore a checkpoint, use [`flax.training.checkpoints.restore_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.restore_checkpoint) and pass in the checkpoint directory. Flax will automatically select the latest checkpoint in the directory. You can also choose to specify a step number or the path of the checkpoint file.

> With the Orbax migration on progress, `restore_checkpoint` can automatically identify whether a checkpoint is saved in legacy or orbax version, and restore the pytree correctly.

Reviewer: Suggestion:

```suggestion
With the migration to Orbax in progress, `restore_checkpoint` can automatically identify whether a checkpoint is saved in the legacy (Flax) or Orbax version, and restore the Pytree correctly.
```

> You can always restore a pytree out of your checkpoints by setting `target=None`.

Reviewer: ```suggestion
You can always restore a Pytree out of your checkpoints by setting `target=None`.
```

(Maybe it's best to double check the spelling on JAX's site. Sometimes it's "PyTree" :thinking: )

Author: JAX site mostly uses "pytree": https://jax.readthedocs.io/en/latest/search.html?q=PyTree# 
Making it consistent here.

Reviewer: Excellent, thanks!

To resolve this, you should pass an example `target` in `flax.training.checkpoints.restore_checkpoint` to let Flax know exactly what structure it should restore to. The `target` should introduce any custom Flax dataclasses explicitly, and have the same structure as the saved checkpoint.

It's often recommended to refactor out the process of initializing a checkpoint's structure (for example, a [`TrainState`](https://flax.readthedocs.io/en/latest/flip/1009-optimizer-api.html?#train-state)), so that saving/loading is easier and less error-prone. This is because complicated objects like `apply_fn` and `tx` (optimizer) are not stored in the checkpoint file and must be initiated by code.

```python
empty_state = train_state.TrainState.create(
    apply_fn=model.apply,
    params=jax.tree_map(np.zeros_like, variables['params']),  # values of the tree leaf doesn't matter
    tx=tx,
)
empty_config = {'dimensions': np.array([0, 0]), 'name': ''}
target = {'model': empty_state, 'config': empty_config, 'data': [jnp.zeros_like(x1)]}
state_restored = checkpoints.restore_checkpoint(ckpt_dir, target=target, step=0)
state_restored
```

### Backward/forward dataclass compatibility

> The flexibility of using *Flax dataclasses*—[`flax.struct.dataclass`](https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html#flax.struct.dataclass)—means that changes in Flax dataclass fields may break your existing checkpoints. For example, if you decide to add a field `batch_stats` to your `TrainState` (like when using [batch normalization](https://flax.readthedocs.io/en/latest/guides/batch_norm.html)), old checkpoints without this field may not be successfully restored. Same goes for removing a field in your dataclass.

Reviewer: fyi @cgarciae @marcvanzee we (@IvyZX and I) decided to reference the BatchNorm guide here :+1:)

Checkpointing is I/O heavy, and if you have a large amount of data to save, it may be worthwhile to put it into a background thread, while continuing with your training.

You can do this by creating an `orbax.AsyncCheckpointer` (as demonstrated in the code cell below) and let it track your save thread.

Note: You should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This enables bookkeeping, such as `keep` and `overwrite` to be consistent across steps.

> Whenever you want to explicitly wait until an async save is done, you can call `async_checkpointer.wait_until_finished()`. Alternatively, you can pass in `orbax_checkpointer=async_checkpointer` when running `restore_checkpoint` and Flax will automatically wait and restore safely.

Reviewer: Suggestion: Add a link to the API docs for `async_checkpointer.wait_until_finished()`.

# for how to set it up in multi-process scenarios.
jax.distributed.initialize("localhost:8889", num_processes=1, process_id=0)

async_checkpointer = orbax.AsyncCheckpointer(orbax.PyTreeCheckpointHandler(), timeout_secs=50)

# Mimic a training loop here:
for step in range(2, 3):
    checkpoints.save_checkpoint(ckpt_dir, ckpt, step=2, overwrite=True, keep=3,
                                orbax_checkpointer=async_checkpointer)
    # ... Continue with your work...

# ... Until a time when you want to wait until the save completes:
async_checkpointer.wait_until_finished()  # Blocks until the checkpoint saving is completed.
checkpoints.restore_checkpoint(ckpt_dir, target=None, step=2)
```

## Multi-host/multi-process checkpointing

JAX provides a few ways to scale up your code on multiple hosts at the same time. This usually happens when the number of devices (CPU/GPU/TPU) is so large that different devices are managed by different hosts (CPU). To get started on JAX in multi-process settings, check out [Using JAX in multi-host and multi-process environments](https://jax.readthedocs.io/en/latest/multi_process.html).

> In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data array needs a special experimental JAX API — [`gda_serialization`](https://github.com/google/jax/blob/c7dcd0913bc8b4878a7e45184553c331254b801a/jax/experimental/gda_serialization/serialization.py) — to save and restore checkpoints. This API lets each host dump its data shards to a single shared storage, such as a Google Cloud bucket.

Reviewer: Maybe we should rename `gda_serialization` -> `array_serialization` (Reference: @yashk2810 's PR `jax.Array` github.com/google/flax/pull/2703)

> Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, it's recommended to use the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can perform computation alongside the saves. 

Reviewer: Double check if it's "Pytree" or "PyTree" (I think it's the former).

To save multi-process arrays, use [`flax.training.checkpoints.save_checkpoint_multiprocess()`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess) in place of `save_checkpoint()` and with the same arguments.

> Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. Please simply use the following code as a sample to run your multi-host checkpointing.

Reviewer: Suggestion:

```suggestion
Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. You can treat the following code as an example for running your multi-host checkpointing:
```

Checkpointing is I/O heavy, and if you have a large amount of data to save, it may be worthwhile to put it into a background thread, while continuing with your training.

You can do this by creating an `orbax.AsyncCheckpointer` (as demonstrated in the code cell below) and let it track your save thread.

> Note that you should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This allows bookkeeping like `keep` and `overwrite` to be consistent across steps.

Reviewer: Suggestion:

```suggestion
Note: You should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This enables bookkeeping, such as `keep` and `overwrite` to be consistent across steps.
```

```

## Multi-host/multi-process checkpointing

JAX provides a few ways to scale up your code on multiple hosts at the same time. This usually happens when the number of devices (CPU/GPU/TPU) is so large that different devices are managed by different hosts (CPU). To get started on JAX in multi-process settings, check out [Using JAX in multi-host and multi-process environments](https://jax.readthedocs.io/en/latest/multi_process.html).

In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data array needs a special experimental JAX API—[`gda_serialization`](https://github.com/google/jax/blob/c7dcd0913bc8b4878a7e45184553c331254b801a/jax/experimental/gda_serialization/serialization.py)—to save and restore checkpoints. This API lets each host dump its data shards to a single shared storage, such as a Google Cloud bucket.

> Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, we recommend using the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can do computation along side the saves. 

Reviewer: Nit:

```suggestion
Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, it's recommended using the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can perform computation alongside the saves. 
```

(First person like "we" and "our" -> second person https://developers.google.com/style/person though "we" sounds OK, IMHO)

> Code wise, just use [`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess) with the same arguments.

Reviewer: Nit: What does "code-wise" mean here? Maybe we can say:

```
In your code, use `flax.training.checkpoints.save_checkpoint_multiprocess`...
...
```

Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. Please simply use the following code as a sample to run your multi-host checkpointing.

```python

> # Multihost related imports.

Reviewer: Nit: "Multi-host" ?
